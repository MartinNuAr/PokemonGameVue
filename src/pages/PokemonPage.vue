<template>
    <h1 v-if="!pokemon">Cargando...</h1>
    <div v-else>

        <h1>¿Quién es este Pokémon?</h1>

        <PokemonPicture 
            :pokemonId="pokemon.id" 
            :showPokemon="showPokemon" 
        />

        <PokemonOptions 
            :pokemonGroup="pokemonArr" 
            @selection="checkAnswer" 
        />

        <div v-if="showAnswer">
            <h2 class="fade-in">{{ message }}</h2>
            <button @click="newGame">
                Nuevo juego
            </button>
        </div>

    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components: {
        PokemonPicture,
        PokemonOptions
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions()

            const randInt = Math.floor( Math.random() * 4)
            this.pokemon = this.pokemonArr[ randInt ]
        },
        checkAnswer(selectedId) {
            this.showPokemon = true
            this.showAnswer = true

            if (selectedId === this.pokemon.id) {
                this.message = `¡Correcto, es ${ this.pokemon.name.charAt(0).toUpperCase() + this.pokemon.name.slice(1) }!`
            } else {
                this.message = `Incorrecto, el Pokémon es ${ this.pokemon.name.charAt(0).toUpperCase() + this.pokemon.name.slice(1) }`
            }
        },
        newGame() {
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null

            this.mixPokemonArray()
        }
    },
    mounted() {
        this.mixPokemonArray()
    }
}
</script>

<style>

</style>