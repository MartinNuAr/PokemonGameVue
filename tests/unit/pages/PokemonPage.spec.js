import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage"
import { pokemon } from "../mocks/pokemon.mock"
import PokemonOptions from "@/components/PokemonOptions"
import PokemonPicture from "@/components/PokemonPicture"

describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('Debe hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de llamar mixPokemonArray al montar', () => {

        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()

    })

    test('Debe hacer match con el snapshot cuando cargan los Pokémon', () => {
        
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemon,
                    pokemon: pokemon[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemon,
                    pokemon: pokemon[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        const pokemonOptions = wrapper.findComponent(PokemonOptions)
        const pokemonPicture = wrapper.findComponent(PokemonPicture)

        expect(pokemonOptions.exists()).toBeTruthy()
        expect(pokemonPicture.exists()).toBeTruthy()

        expect(pokemonPicture.attributes('pokemonid')).toBe("5")
        expect(pokemonOptions.attributes('pokemongroup')).toBeTruthy()

    })

    test('Pruebas con checkAnswer', async() => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemon,
                    pokemon: pokemon[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(5)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`¡Correcto, es ${ pokemon[0].name.charAt(0).toUpperCase() + pokemon[0].name.slice(1) }!`)

        await wrapper.vm.checkAnswer(10)
        
        expect(wrapper.vm.message).toBe(`Incorrecto, el Pokémon es ${ pokemon[0].name.charAt(0).toUpperCase() + pokemon[0].name.slice(1) }`)
    })

})