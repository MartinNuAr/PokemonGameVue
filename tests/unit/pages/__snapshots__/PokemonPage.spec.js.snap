// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe hacer match con el snapshot 1`] = `<h1>Cargando...</h1>`;

exports[`PokemonPage Component Debe hacer match con el snapshot cuando cargan los Pokémon 1`] = `
<div>
  <h1>¿Quién es este Pokémon?</h1>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemongroup="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
